### PR TITLE
[WIP] Add before/after hooks, refactor globals

### PR DESCRIPTION
**Pull request by yohanboniface**
Allows to do things like:

```
@cli
def mycommand():
    do_something_with_db()

@before
async def initdb(dbname):
    conn = init_the_db()

if __name__ == '__main__':
    run(dbname='defaultname')
```
Then
```
$ ./myscript mycommand --dbname myname
```

### CODE REVIEW

File: minicli/__init__.py [hunk 1]
```diff
@@ -31,31 +27,20 @@ def __call__(self, *args, **kwargs):
         except KeyboardInterrupt:
             pass
 
-    def invoke(self, parsed):
+    def invoke(self, parsed, **globals):
```
Comment (davidbgk) [inline minicli/__init__.py]: Overriding the existing `globals` doesn't smell good to me. Proposition: `shared`?

File: minicli/__init__.py [hunk 2]
```diff
@@ -2,14 +2,11 @@
...
 
-
 NO_DEFAULT = inspect._empty
 NARGS = ...
-
-parser = argparse.ArgumentParser()
-subparsers = parser.add_subparsers(title='Available commands', metavar='')
-
-GLOBALS = {}
+BEFORE = []
```
Comment (davidbgk) [inline minicli/__init__.py]: Why uppercase if not constants?

File: minicli/__init__.py [hunk 3]
```diff
@@ -143,21 +111,90 @@ def cli(*args, **kwargs):
...
+    BEFORE.append(func)
+    return func
+
+
+def after(func):
+    AFTER.append(func)
+    return func
+
+
+def after_before(funcs, **globals):
```
Comment (davidbgk) [inline minicli/__init__.py]: I would call this `hooks` and rename `BEFORE` to `pre_hooks` and `AFTER` to `post_hooks`, thoughts?

File: minicli/__init__.py [hunk 4]
```diff
@@ -143,21 +111,90 @@ def cli(*args, **kwargs):
...
+    for func in funcs:
+        spec = inspect.signature(func)
+        async = inspect.iscoroutinefunction(func)
```
Comment (davidbgk) [inline minicli/__init__.py]: 💣 reserved keyword?

File: minicli/__init__.py (hunk 1 above)
Comment (yohanboniface) [reply minicli/__init__.py]: Sounds good, fixed :)

File: minicli/__init__.py (hunk 3 above)
Comment (yohanboniface) [reply minicli/__init__.py]: Done!

File: minicli/__init__.py (hunk 4 above)
Comment (yohanboniface) [reply minicli/__init__.py]: Uh! Fun that it worked. Fixed anyway :)

Comment (davidbgk) [reply minicli/__init__.py]: I was surprised too, especially with `if async` ;)

**Pull request merged**